CONTRIBUTING.md

```diff
@@ -0,0 +1,223 @@
+
+The majority of our types are defined in the `domain` layer (see Architecture section below). We generally create types for the following scenarios:
+
+#### Symbol types
+These are unique types that are alternatives to generically typing things as Typescript's primitive types (e.g. `string`, `number`, `boolean`). Doing this helps us to add context to primitive types that we pass around and it allows the type checked to distinguish between different kinds of primitives throughout the code.
+
+For example, an onchain address and a lightning network payment request are both strings, but they aren't interchangeable as a data type. Instead of using `string` type for these types we would define as follows using a "unique symbol" trick:
+
+```
+declare const encodedPaymentRequestSymbol: unique symbol
+type EncodedPaymentRequest = string & { [encodedPaymentRequestSymbol]: never }
+
+declare const onChainAddressSymbol: unique symbol
+type OnChainAddress = string & { [onChainAddressSymbol]: never }
+
+```
+
+#### Error types
+
+These are types mostly used in function signature type definitions. They derive from implemented error classes and need to be imported in a special way to their type declaration file.
+
+For example, an error may be defined in a module's `error.ts` file like this:
+```
+export class LightningError extends Error {
+  name = this.constructor.name
+}
+```
+
+and then it is imported to its `index.types.d.ts` declaration file and made into a type like this:
+```
+type LightningError = import("./errors").LightningError
+```
+
+
+#### Imported Library types
+
+In the places where we would like to work with types defined in imported libraries, we have two options. When we would like to use them directly, we can simply import them. If we would like to re-use a type to create our own types, we would re-import to get those types into our declaration files.
+
+For example, for a result type from the `lightning` library, we would import and re-use it like:
+```
+type GetPaymentResult = import("lightning").GetPaymentResult
+type RawPaths = NonNullable<GetPaymentResult["payment"]>["paths"]
+```
+
+In this example, we need the type definition for the `payment.paths` property of the `GetPaymentResult` type from the library that we otherwise would not have direct access to.
+
+
+#### Function signatures and argument types
+Whenever we implement a new function, the argument and return types are assigned at the point of the function's implementation.
+
+For example:
+```ts
+const myFunction = async (myArg: MyArg): ThingResult => {
+    return doThing(myArg)
+}
+
+const myOtherFunction = async ({
+  myFirstArg,
+  mySecondArg,
+}: {
+  myFirstArg: MyFirstArg
+  mySecondArg: MySecondArg
+}): ThingResult => {
```

Comment:
if the function is async should probable be `Promise<ThingResult>` for the type in return